src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs

```diff
@@ -271,33 +274,35 @@ private static void CheckDefaultProvider(Type type)
                     return;
                 }
 
-                // Immediately clear this. If we find a default provider
-                // and it starts messing around with type information,
+                // Immediately set this so we can short circuit in the ContainsKey check above.
+                // If we find a default provider and it starts messing around with type information,
                 // this could infinitely recurse.
                 s_defaultProviders[type] = null;
-            }
 
-            // Always use core reflection when checking for
-            // the default provider attribute. If there is a
-            // provider, we probably don't want to build up our
-            // own cache state against the type. There shouldn't be
-            // more than one of these, but walk anyway. Walk in
-            // reverse order so that the most derived takes precidence.
-            object[] attrs = type.GetCustomAttributes(typeof(TypeDescriptionProviderAttribute), false);
-            bool providerAdded = false;
-            for (int idx = attrs.Length - 1; idx >= 0; idx--)
-            {
-                TypeDescriptionProviderAttribute pa = (TypeDescriptionProviderAttribute)attrs[idx];
-                Type? providerType = Type.GetType(pa.TypeName);
-                if (providerType != null && typeof(TypeDescriptionProvider).IsAssignableFrom(providerType))
-                {
-                    TypeDescriptionProvider prov = (TypeDescriptionProvider)Activator.CreateInstance(providerType)!;
-                    AddProvider(prov, type);
-                    providerAdded = true;
+                // Always use core reflection when checking for
+                // the default provider attribute. If there is a
+                // provider, we probably don't want to build up our
+                // own cache state against the type. There shouldn't be
+                // more than one of these, but walk anyway. Walk in
+                // reverse order so that the most derived takes precedence.
+                object[] attrs = type.GetCustomAttributes(typeof(TypeDescriptionProviderAttribute), false);
+                for (int idx = attrs.Length - 1; idx >= 0; idx--)
+                {
+                    TypeDescriptionProviderAttribute pa = (TypeDescriptionProviderAttribute)attrs[idx];
+                    Type? providerType = Type.GetType(pa.TypeName);
+                    if (providerType != null && typeof(TypeDescriptionProvider).IsAssignableFrom(providerType))
+                    {
+                        TypeDescriptionProvider prov = (TypeDescriptionProvider)Activator.CreateInstance(providerType)!;
+                        AddProvider(prov, type);
+                        providerAdded = true;
+                    }
                 }
+
+                s_defaultProviders[type] = s_initializedProvider;
```

Comment:
I _think_ this needs to happen after the call to `CheckDefaultProvider(baseType);` and that needs to remain under the lock.  Otherwise another thread might exit out of this method before the default provider is added for the base type.

To test that you could try testing this with a race where the converter is not set on the target type, but it's base type.

Reply:
Agree - expanded the lock. Also for the base type check, split the method into 2 to prevent an unnecessary dictionary check and a recursive lock.

Reply:
Makes sense - that definitely makes this clearer